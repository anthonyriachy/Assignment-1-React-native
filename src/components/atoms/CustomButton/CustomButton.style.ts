import { StyleSheet } from 'react-native';
import { colors } from '../../../constants/colors';
import { globalStyles } from '../../../constants/globalStyles';

export const styles = StyleSheet.create({
    button:{
        backgroundColor:colors.primary,
        padding: 10,
        borderRadius: 10,
        alignItems: 'center',
        justifyContent:'center',
    },
    text: {
        color: 'white',
        fontSize: 15,
        fontFamily:globalStyles.semiBold,
    },
});
