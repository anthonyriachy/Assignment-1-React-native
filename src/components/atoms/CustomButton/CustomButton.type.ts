import { ButtonProps, StyleProp, ViewStyle } from 'react-native';

export type CustomButtonProps = {
    disabled?: boolean;
    loading?: boolean;
    style?: StyleProp<ViewStyle>;
}&ButtonProps;
