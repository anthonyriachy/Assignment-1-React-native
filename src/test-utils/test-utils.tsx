import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from '../context/ThemeContext/ThemeContext';

// Mock theme hook
jest.mock('../hooks/UseTheme', () => ({
  useTheme: () => ({
    colors: {
      background: '#fff',
      text: '#000',
      primary: '#007AFF',
      border: '#ccc',
    },
    theme: 'light',
  }),
}));

// Custom render function that includes providers
const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: ThemeProvider, ...options });

// Re-export everything
export * from '@testing-library/react-native';

// Override render method
export { customRender as render }; 