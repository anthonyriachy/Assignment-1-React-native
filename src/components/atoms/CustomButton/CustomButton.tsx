import React from 'react';
import {TouchableOpacity, Text} from 'react-native';
import { CustomButtonProps } from './CustomButton.type';
import { styles } from './CustomButton.style';
import { CustomText } from '../CustomText/CustomText';

export const CustomButton = ({title,onPress,disabled,loading,style:customStyle}:CustomButtonProps)=>{
    return(
        <TouchableOpacity
            testID="custom-button"
            style={[styles.button,customStyle]}
            onPress={onPress}
            disabled={disabled || loading}
        >
            <CustomText style={styles.text}>{loading ? title+'...' : title}</CustomText>
        </TouchableOpacity>
    );
};
