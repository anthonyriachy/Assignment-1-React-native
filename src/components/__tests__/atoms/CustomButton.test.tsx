import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { render } from '../../../../src/test-utils/test-utils';
import { CustomButton } from '../../atoms/CustomButton/CustomButton';

// Mock child components
jest.mock('../../atoms/CustomText/CustomText', () => {
  const React = require('react');
  const { Text } = require('react-native');
  return {
    CustomText: ({ children, style }: any) => (
      <Text style={style}>{children}</Text>
    ),
  };
});

describe('CustomButton', () => {
  const mockOnPress = jest.fn();

  beforeEach(() => {
    mockOnPress.mockClear();
  });

  it('renders correctly with title', () => {
    const { getByText } = render(
      <CustomButton title="Test Button" onPress={mockOnPress} />
    );

    expect(getByText('Test Button')).toBeTruthy();
  });

  it('calls onPress when pressed', () => {
    const { getByText } = render(
      <CustomButton title="Test Button" onPress={mockOnPress} />
    );

    fireEvent.press(getByText('Test Button'));
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it('shows loading state correctly', () => {
    const { getByText } = render(
      <CustomButton title="Test Button" onPress={mockOnPress} loading={true} />
    );

    expect(getByText('Test Button...')).toBeTruthy();
  });

  it('is disabled when disabled prop is true', () => {
    const { getByText } = render(
      <CustomButton title="Test Button" onPress={mockOnPress} disabled={true} />
    );

    const button = getByText('Test Button');
    fireEvent.press(button);
    expect(mockOnPress).not.toHaveBeenCalled();
  });

  it('is disabled when loading prop is true', () => {
    const { getByText } = render(
      <CustomButton title="Test Button" onPress={mockOnPress} loading={true} />
    );

    const button = getByText('Test Button...');
    fireEvent.press(button);
    expect(mockOnPress).not.toHaveBeenCalled();
  });

  it('applies custom styles', () => {
    const customStyle = { backgroundColor: 'red' };
    const { getByTestId } = render(
      <CustomButton 
        title="Test Button" 
        onPress={mockOnPress} 
        style={customStyle}
      />
    );

    const button = getByTestId('custom-button');
    expect(button.props.style).toMatchObject(customStyle);
  });

  it('renders correctly with default props', () => {
    const { toJSON } = render(
      <CustomButton 
        title="Test Button"
        onPress={jest.fn()}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when disabled', () => {
    const { toJSON } = render(
      <CustomButton 
        title="Disabled Button"
        onPress={jest.fn()}
        disabled={true}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when loading', () => {
    const { toJSON } = render(
      <CustomButton 
        title="Loading Button"
        onPress={jest.fn()}
        loading={true}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
}); 