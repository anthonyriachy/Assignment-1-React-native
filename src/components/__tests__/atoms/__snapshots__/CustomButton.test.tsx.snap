// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomButton renders correctly when disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#6055D8",
      "borderRadius": 10,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
    }
  }
  testID="custom-button"
>
  <Text
    style={
      {
        "color": "white",
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 15,
      }
    }
  >
    Disabled Button
  </Text>
</View>
`;

exports[`CustomButton renders correctly when loading 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#6055D8",
      "borderRadius": 10,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
    }
  }
  testID="custom-button"
>
  <Text
    style={
      {
        "color": "white",
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 15,
      }
    }
  >
    Loading Button...
  </Text>
</View>
`;

exports[`CustomButton renders correctly with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#6055D8",
      "borderRadius": 10,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
    }
  }
  testID="custom-button"
>
  <Text
    style={
      {
        "color": "white",
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 15,
      }
    }
  >
    Test Button
  </Text>
</View>
`;
